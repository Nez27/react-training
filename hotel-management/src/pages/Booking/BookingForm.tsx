import { useCallback, useMemo } from 'react';
import { FormProvider, useForm } from 'react-hook-form';

// Styled
import Input from '@commonStyle/Input.ts';

// Constants
import { REQUIRED_FIELD_ERROR } from '@constant/formValidateMessage.ts';

// Components
import Form from '@component/Form/index.tsx';
import Select, { ISelectOptions } from '@component/Select/index.tsx';

// Hooks
import { useCreateBooking } from '@hook/bookings/useCreateBooking.ts';
import { useUpdateBooking } from '@hook/bookings/useUpdateBooking.ts';
import { useUserRoomAvailable } from '@hook/useUserRoomAvailable.ts';

// Helpers
import {
  convertCurrencyToNumber,
  formatCurrency,
  getDayDiff,
} from '@helper/helper.ts';

// Services
import { getRoomById, updateRoomStatus } from '@service/roomServices.ts';
import { updateUserBookedStatus } from '@service/userServices.ts';

// Types
import { IBooking, TBookingResponse } from '@type/booking.ts';

interface IBookingFormProp {
  onCloseModal?: () => void;
  booking?: TBookingResponse;
}

interface IBookingForm extends Omit<IBooking, 'amount'> {
  amount: string;
}

const BookingForm = ({ onCloseModal, booking }: IBookingFormProp) => {
  const { roomsAvailable, usersAvailable, dispatch } = useUserRoomAvailable();
  const { isCreating, createBooking } = useCreateBooking();
  const { isUpdating, updateBooking } = useUpdateBooking();
  const isLoading = isCreating || isUpdating;
  const { id: editId, ...editValues } = { ...booking };
  const formMethods = useForm<IBookingForm>({
    defaultValues: editId
      ? {
          startDate: editValues.startDate,
          endDate: editValues.endDate,
          roomId: editValues.rooms?.id,
          userId: editValues.users?.id,
          amount: formatCurrency(editValues.amount!),
        }
      : {},
  });
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isValid },
    trigger,
    getValues,
    setValue,
  } = formMethods;

  // Submit form
  const onSubmit = async (newBooking: IBookingForm) => {
    if (!editId) {
      // Add request
      createBooking(
        { ...newBooking, amount: convertCurrencyToNumber(newBooking.amount) },
        {
          onSuccess: async () => {
            reset();
            onCloseModal?.();

            // Update room, user available in global state
            dispatch!({
              type: 'updateStatusUser',
              payload: [{ id: newBooking.userId, isBooked: true }],
            });
            dispatch!({
              type: 'updateStatusRoom',
              payload: [{ id: newBooking.roomId, status: true }],
            });

            // Update room, user available in server
            await updateRoomStatus(newBooking.roomId, true);
            await updateUserBookedStatus(newBooking.userId, true);
          },
        }
      );
    } else {
      // Edit request
      newBooking.id = editId!;
      updateBooking(
        { ...newBooking, amount: convertCurrencyToNumber(newBooking.amount) },
        {
          onSuccess: async () => {
            reset();
            onCloseModal?.();

            // Update room status in global state
            // Reset status old room
            dispatch!({
              type: 'updateStatusRoom',
              payload: [{ id: booking!.rooms!.id, status: false }],
            });

            // Change status new room
            dispatch!({
              type: 'updateStatusRoom',
              payload: [{ id: newBooking.roomId, status: true }],
            });

            // Update room status in server
            // Reset status old room
            await updateRoomStatus(booking!.rooms!.id, false);

            // Change status new room
            await updateRoomStatus(newBooking.roomId, true);
          },
        }
      );
    }
  };

  // Init user, room available to options choice
  const userOptions = useMemo(() => {
    const options: ISelectOptions[] = [];

    usersAvailable?.forEach((item) => {
      if (booking?.users?.id === item.id || !item.isBooked) {
        options.push({
          label: item.name!,
          value: item.id.toString(),
        });
      }
    });

    return options;
  }, [usersAvailable, booking?.users?.id]);

  const roomOptions = useMemo(() => {
    const options: ISelectOptions[] = [];

    roomsAvailable?.forEach((item) => {
      if (booking?.rooms?.id === item.id || !item.status) {
        options.push({
          label: item.name!,
          value: item.id.toString(),
        });
      }
    });

    return options;
  }, [roomsAvailable, booking?.rooms?.id]);

  // Calculate the final price
  const computePrice = useCallback(async () => {
    setValue('amount', 'Loading...');
    const startDateValue = getValues('startDate');
    const endDateValue = getValues('endDate');
    const roomValue = getValues('roomId');

    if (startDateValue && endDateValue && roomValue) {
      const daysDiff = getDayDiff(
        new Date(startDateValue),
        new Date(endDateValue)
      );

      // Fetch room data
      const room = await getRoomById(roomValue.toString());
      const amount = daysDiff * room.price;

      setValue('amount', formatCurrency(amount), { shouldValidate: true });
    }
  }, [getValues, setValue]);

  const startDateValidate = useMemo(
    () => new Date().toISOString().split('T')[0],
    []
  );

  const endDateValidate = () => {
    if (getValues('startDate')) {
      const date = new Date(getValues('startDate'));
      date.setDate(date.getDate() + 1);

      return date.toISOString().split('T')[0];
    }

    return;
  };

  return (
    <FormProvider {...formMethods}>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Row label="User" error={errors?.userId?.message}>
          {userOptions.length ? (
            <Select
              ariaLabel="user-select"
              options={userOptions}
              id="userId"
              disable={Boolean(booking?.users)}
              optionsConfigForm={{
                valueAsNumber: true,
                onChange: () => trigger('userId'),
              }}
            />
          ) : (
            <p>No user available</p>
          )}
        </Form.Row>

        <Form.Row label="Room" error={errors?.roomId?.message}>
          {roomOptions.length ? (
            <Select
              ariaLabel="user-select"
              options={roomOptions}
              id="roomId"
              optionsConfigForm={{
                valueAsNumber: true,
                onChange: () => trigger('roomId'),
              }}
            />
          ) : (
            <p>No room available</p>
          )}
        </Form.Row>

        <Form.Row label="Start Date" error={errors?.startDate?.message}>
          <Input
            type="date"
            id="startDate"
            min={startDateValidate}
            {...register('startDate', {
              required: REQUIRED_FIELD_ERROR,
              onChange: () => {
                trigger('startDate');
                computePrice();
              },
            })}
          />
        </Form.Row>

        <Form.Row label="End Date" error={errors?.startDate?.message}>
          <Input
            type="date"
            id="endDate"
            min={endDateValidate()}
            disabled={!getValues('startDate')}
            {...register('endDate', {
              required: REQUIRED_FIELD_ERROR,
              onChange: () => {
                trigger('endDate');
                computePrice();
              },
            })}
          />
        </Form.Row>

        <Form.Row label="Amount">
          <Input
            type="text"
            id="amount"
            {...register('amount', {
              required: REQUIRED_FIELD_ERROR,
              onChange: () => trigger('amount'),
            })}
            readOnly
          />
        </Form.Row>

        <Form.Action>
          <Form.Button
            type="submit"
            name="submit"
            disabled={!isDirty || !isValid || isLoading}
          >
            {
              !editId 
                ? 'Add' 
                : 'Save'
            }
          </Form.Button>
          <Form.Button type="button" variations="secondary" onClick={onCloseModal}>
            Close
          </Form.Button>
        </Form.Action>
      </Form>
    </FormProvider>
  );
};

export default BookingForm;
