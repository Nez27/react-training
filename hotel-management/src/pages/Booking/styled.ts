import styled from 'styled-components';

// Components
import Button from '@commonStyle/Button';

const StyledBooking = styled.main`
  padding: 20px;
  padding-bottom: 100px;

  display: flex;
  flex-direction: column;
  gap: 30px;
`;

const Title = styled.h2`
  font-size: var(--fs-md);
  color: var(--dark-text);
  text-transform: capitalize;
`;

const FormBtn = styled(Button)`
  width: 100%;

  &:disabled,
  &[disabled] {
    background-color: var(--disabled-btn-color);
  }
`;

const StyledOperationTable = styled.div`
  display: flex;
  gap: 30px;
  justify-content: flex-end;
`;

export { StyledBooking, Title, StyledOperationTable, FormBtn };
