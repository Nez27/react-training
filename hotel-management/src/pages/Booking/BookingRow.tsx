import { useCallback, useMemo } from 'react';

// Helpers
import { formatCurrency } from '@helper/helper';

// Types
import { TBookingResponse } from '@type/booking';

// Components
import Modal from '@component/Modal';
import Table from '@component/Table';
import Menus from '@component/Menus';
import { RiEditBoxFill } from 'react-icons/ri';
import { TbArrowNarrowRight } from 'react-icons/tb';

// Hooks

// Constants
import { FORM } from '@constant/commons';
import BookingForm from './BookingForm';

interface IBookingRow {
  booking: TBookingResponse;
}

const BookingRow = ({ booking }: IBookingRow) => {
  const { id, users, startDate, endDate, rooms, amount, status } = booking;
  // const { isDeleting, deleteRoom } = useDeleteBooking();
  const statusText = status ? 'Check in' : 'Check out';
  const formattedPrice = useMemo(() => formatCurrency(amount), [amount]);
  const renderEditBtn = useCallback(
    (onCloseModal: () => void) => (
      <Menus.Button onClick={onCloseModal} icon={<RiEditBoxFill />}>
        Edit
      </Menus.Button>
    ),
    []
  );

  return (
    <Table.Row>
      <div>{users?.name}</div>
      <div style={{ display: 'flex', alignItems: 'center' }}>
        {startDate} &nbsp; <TbArrowNarrowRight />
        &nbsp; {endDate}
      </div>
      <div>{rooms?.name}</div>
      <div>{formattedPrice}</div>
      <div>{statusText}</div>

      <div>
        <Modal>
          <Menus.Menu>
            <Menus.Toggle id={id.toString()} />

            <Menus.List id={id.toString()}>
              <Modal.Open
                modalName={FORM.EDIT}
                renderChildren={renderEditBtn}
              />
            </Menus.List>

            <Modal.Window name={FORM.EDIT} title="Edit Booking">
              <BookingForm booking={booking} key={booking.id} />
            </Modal.Window>
          </Menus.Menu>
        </Modal>
      </div>
    </Table.Row>
  );
};

export default BookingRow;
