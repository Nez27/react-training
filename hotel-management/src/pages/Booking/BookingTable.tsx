import { useCallback } from 'react';

// Components
import Menus from '@component/Menus';
import Table from '@component/Table';
import Message from '@component/Message';
import Search from '@component/Search';
import SortBy from '@component/SortBy';
import OrderBy from '@component/OrderBy';

// Types

// Constants
import { ORDERBY_OPTIONS, ROOM_PAGE } from '@constant/commons';

// Styled
import { StyledOperationTable } from './styled';
import Direction from '@commonStyle/Direction';
import Spinner from '@commonStyle/Spinner';

// Hooks
import Pagination from '@component/Pagination';
import { useBookings } from '@hook/bookings/useBookings';
import { TBookingResponse } from '@type/booking';
import BookingRow from './BookingRow';

const BookingTable = () => {
  const columnName = ['User', 'Date', 'Room', 'Amount', 'Status'];
  const { isLoading, bookings, count } = useBookings();

  const renderBookingRow = useCallback(
    (booking: TBookingResponse) => (
      <BookingRow booking={booking} key={booking.id} />
    ),
    []
  );

  return (
    <>
      <Direction>
        <StyledOperationTable>
          <OrderBy options={ORDERBY_OPTIONS} />

          <SortBy options={ROOM_PAGE.SORTBY_OPTIONS} />
          <Search setPlaceHolder="Search by name..." />
        </StyledOperationTable>

        {isLoading && <Spinner />}

        {bookings && bookings.length ? (
          <Menus>
            <Table columns="15% 25% 20% 15% 10% 10% 5%">
              <Table.Header headerColumn={columnName} />
              <Table.Body<TBookingResponse>
                data={bookings}
                render={renderBookingRow}
              />
              <Table.Footer>
                <Pagination count={count!} />
              </Table.Footer>
            </Table>
          </Menus>
        ) : (
          !isLoading && <Message>No data to show here!</Message>
        )}
      </Direction>
    </>
  );
};

export default BookingTable;
