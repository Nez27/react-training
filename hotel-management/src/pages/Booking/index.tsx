import Direction from '@commonStyle/Direction';
import styled from 'styled-components';
import { Title } from './styled';
import Modal from '@component/Modal';
import Button from '@commonStyle/Button';
import BookingTable from './BookingTable';
import BookingForm from './BookingForm';

// Constants
const bookingForm = 'booking-form';

const StyledBooking = styled.main`
  padding: 20px;
  padding-bottom: 100px;

  display: flex;
  flex-direction: column;
  gap: 30px;
`;

const Booking = () => {

  return (
    <StyledBooking>
      <Direction type="horizontal">
        <Title>List Booking</Title>

        <Modal>
          <Modal.Open
            modalName={bookingForm}
            renderChildren={(onCloseModal) => (
              <Button onClick={onCloseModal}>Add booking</Button>
            )}
          />
          <Modal.Window name={bookingForm} title="Add form">
            <BookingForm />
          </Modal.Window>
        </Modal>
      </Direction>

      <BookingTable />
    </StyledBooking>
  );
};

export default Booking;
