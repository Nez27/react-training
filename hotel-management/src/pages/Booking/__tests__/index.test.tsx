import { BrowserRouter } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';

// Components
import Booking from '..';

describe('Booking', () => {
  const queryClient = new QueryClient();

  const wrapper = render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>
        <Booking />
      </BrowserRouter>
    </QueryClientProvider>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
