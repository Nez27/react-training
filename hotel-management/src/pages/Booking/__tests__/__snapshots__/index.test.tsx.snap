// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="sc-gtsqUy eStAWW"
      >
        <div
          class="sc-bdnyFh bpfZjm"
          type="horizontal"
        >
          <h2
            class="sc-dlniIP kjGsqR"
          >
            List Booking
          </h2>
          <button
            class="sc-eCAqax iNUTro"
            variations="primary"
          >
            Add booking
          </button>
        </div>
        <div
          class="sc-bdnyFh iucOVv"
          type="vertical"
        >
          <div
            class="sc-hKFymg frjaax"
          >
            <input
              class="sc-iemXqs HAzOF"
              placeholder="Search by name..."
            />
          </div>
          <div
            class="sc-bYwyHq csuXEL"
          />
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="sc-gtsqUy eStAWW"
    >
      <div
        class="sc-bdnyFh bpfZjm"
        type="horizontal"
      >
        <h2
          class="sc-dlniIP kjGsqR"
        >
          List Booking
        </h2>
        <button
          class="sc-eCAqax iNUTro"
          variations="primary"
        >
          Add booking
        </button>
      </div>
      <div
        class="sc-bdnyFh iucOVv"
        type="vertical"
      >
        <div
          class="sc-hKFymg frjaax"
        >
          <input
            class="sc-iemXqs HAzOF"
            placeholder="Search by name..."
          />
        </div>
        <div
          class="sc-bYwyHq csuXEL"
        />
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
